File: sdk/python/tempmail_sdk/providers/email10min.py
"""
email10min — https://email10min.com
GET /zh 拿 CSRF + Cookie，POST /messages 获取邮件。
"""

import base64
import json
import re
import time
from typing import List, Optional

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "email10min"
BASE_URL = "https://email10min.com"

BROWSER_HEADERS = {
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
    "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
    "Cache-Control": "no-cache",
    "DNT": "1",
    "Pragma": "no-cache",
    "Upgrade-Insecure-Requests": "1",
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
    ),
    "sec-ch-ua": '"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"',
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": '"Windows"',
}

AJAX_HEADERS = {
    "Accept": "application/json, text/plain, */*",
    "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
    "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
    "Origin": BASE_URL,
    "Referer": f"{BASE_URL}/zh",
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
    ),
    "X-Requested-With": "XMLHttpRequest",
    "sec-ch-ua": '"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"',
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": '"Windows"',
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-origin",
}

TOK_PREFIX = "e10m:"

_csrf_meta_re = re.compile(r'<meta\s+name="csrf-token"\s+content="([^"]+)"', re.I)
_csrf_input_re = re.compile(r'<input[^>]+name="_token"[^>]+value="([^"]+)"', re.I)
_email_id_re = re.compile(r'id="emailAddress"[^>]*>([^<]+)', re.I)
_email_cls_re = re.compile(r'class="[^"]*email[^"]*"[^>]*>([^<]*@[^<]+)', re.I)
_email_data_re = re.compile(r'data-email="([^"]+@[^"]+)"', re.I)
_email_value_re = re.compile(
    r'value="([a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,})"'
)
_email_json_re = re.compile(r'"mailbox"\s*:\s*"([^"]+@[^"]+)"')
_email_generic_re = re.compile(r"([a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,})")


def _cookie_header_from_response(r) -> str:
    """从响应中提取 Cookie 字符串"""
    parts = []
    for c in r.cookies:
        parts.append(f"{c.name}={c.value}")
    return "; ".join(parts)


def _encode_token(cookie: str, csrf: str) -> str:
    raw = json.dumps({"c": cookie, "t": csrf})
    return TOK_PREFIX + base64.b64encode(raw.encode("utf-8")).decode("utf-8")


def _decode_token(token: str) -> tuple:
    if not token.startswith(TOK_PREFIX):
        raise ValueError("email10min: invalid session token")
    try:
        raw = base64.b64decode(token[len(TOK_PREFIX) :]).decode("utf-8")
    except Exception:
        raise ValueError("email10min: invalid session token")
    data = json.loads(raw)
    cookie = (data.get("c") or "").strip()
    csrf = (data.get("t") or "").strip()
    if not cookie or not csrf:
        raise ValueError("email10min: invalid session token (empty fields)")
    return cookie, csrf


def _extract_csrf(html: str) -> str:
    m = _csrf_meta_re.search(html)
    if m:
        return m.group(1)
    m = _csrf_input_re.search(html)
    if m:
        return m.group(1)
    raise RuntimeError("email10min: 未找到 CSRF token")


def _extract_email(html: str) -> str:
    m = _email_id_re.search(html)
    if m and "@" in m.group(1):
        return m.group(1).strip()
    m = _email_cls_re.search(html)
    if m:
        return m.group(1).strip()
    m = _email_data_re.search(html)
    if m:
        return m.group(1).strip()
    m = _email_value_re.search(html)
    if m:
        return m.group(1).strip()
    m = _email_json_re.search(html)
    if m:
        return m.group(1).strip()
    m = _email_generic_re.search(html)
    if m:
        addr = m.group(1)
        if "email10min" not in addr and "example" not in addr and "google" not in addr:
            return addr.strip()
    raise RuntimeError("email10min: 未找到邮箱地址")


def generate_email() -> EmailInfo:
    """创建 email10min 临时邮箱"""
    r = tm_http.get(f"{BASE_URL}/zh", headers=BROWSER_HEADERS, timeout=15)
    r.raise_for_status()

    cookie = _cookie_header_from_response(r)
    html = r.text

    csrf = _extract_csrf(html)
    email_addr = _extract_email(html)
    token = _encode_token(cookie, csrf)

    return EmailInfo(channel=CHANNEL, email=email_addr, _token=token)


def get_emails(email: str, token: str) -> List[Email]:
    """获取 email10min 邮件列表"""
    cookie, csrf = _decode_token(token)
    ts = str(int(time.time() * 1000))

    h = dict(AJAX_HEADERS)
    h["Cookie"] = cookie

    r = tm_http.post(
        f"{BASE_URL}/messages?{ts}",
        headers=h,
        data=f"_token={csrf}&captcha=",
        timeout=15,
    )
    r.raise_for_status()

    data = r.json()
    messages = data.get("messages") or []
    if not isinstance(messages, list):
        return []

    out: List[Email] = []
    for i, raw in enumerate(messages):
        if not isinstance(raw, dict):
            continue
        msg_id = raw.get("id") or raw.get("message_id") or str(i)
        from_addr = raw.get("from") or raw.get("sender") or ""
        to_addr = raw.get("to") or email
        subject = raw.get("subject") or ""
        text = raw.get("text") or raw.get("body") or ""
        html_content = raw.get("html") or raw.get("body_html") or ""
        date = raw.get("date") or raw.get("created_at") or ""

        flat = {
            "id": str(msg_id),
            "from": from_addr,
            "to": to_addr,
            "subject": subject,
            "text": text,
            "html": html_content,
            "date": date,
            "isRead": False,
        }
        out.append(normalize_email(flat, email))
    return out

File: sdk/ruby/lib/tempmail_sdk/providers/email10min.rb
# frozen_string_literal: true

require "base64"
require "json"

module TempmailSdk
  module Providers
    # email10min.com 渠道实现
    #
    # 流程：
    #   1. GET /zh 获取 CSRF meta token + session cookies + HTML 中的邮箱地址
    #   2. POST /messages?<ts> 携带 Cookie + _token 表单 获取邮件列表
    # token 格式: "e10m:" + base64(JSON{c: cookie, t: csrf})
    module Email10min
      CHANNEL = "email10min"
      BASE_URL = "https://email10min.com"
      TOK_PREFIX = "e10m:"

      CSRF_META_RE = /<meta\s+name="csrf-token"\s+content="([^"]+)"/.freeze
      CSRF_INPUT_RE = /<input[^>]+name="_token"[^>]+value="([^"]+)"/.freeze
      EMAIL_ID_RE = /id="emailAddress"[^>]*>([^<]+)/.freeze
      EMAIL_DATA_RE = /data-email="([^"]+@[^"]+)"/.freeze
      EMAIL_JSON_RE = /"mailbox"\s*:\s*"([^"]+@[^"]+)"/.freeze
      EMAIL_GENERIC_RE = /([a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,})/.freeze

      BROWSER_HEADERS = {
        "Accept" => "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
        "Accept-Language" => "zh-CN,zh;q=0.9,en;q=0.8",
        "User-Agent" => "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " \
                       "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
      }.freeze

      AJAX_HEADERS = {
        "Accept" => "application/json, text/plain, */*",
        "Accept-Language" => "zh-CN,zh;q=0.9,en;q=0.8",
        "Content-Type" => "application/x-www-form-urlencoded; charset=UTF-8",
        "Origin" => BASE_URL,
        "Referer" => "#{BASE_URL}/zh",
        "User-Agent" => "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " \
                       "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
        "X-Requested-With" => "XMLHttpRequest"
      }.freeze

      module_function

      # 从响应的 Set-Cookie 头合并为 Cookie 字符串
      def merge_cookies(resp)
        parts = []
        resp.set_cookies.each do |line|
          first = line.split(";", 2).first.to_s.strip
          parts << first unless first.empty?
        end
        parts.join("; ")
      end

      # token 编码
      def encode_token(cookie, csrf)
        raw = JSON.generate("c" => cookie, "t" => csrf)
        TOK_PREFIX + Base64.strict_encode64(raw)
      end

      # token 解码，返回 [cookie, csrf]
      def decode_token(token)
        raise "email10min: 无效 session token" unless token.to_s.start_with?(TOK_PREFIX)

        raw = Base64.strict_decode64(token[TOK_PREFIX.length..])
        data = JSON.parse(raw)
        cookie = data["c"].to_s
        csrf = data["t"].to_s
        raise "email10min: 无效 session token（字段缺失）" if cookie.empty? || csrf.empty?

        [cookie, csrf]
      rescue ArgumentError, JSON::ParserError
        raise "email10min: 无效 session token"
      end

      # 从 HTML 中提取 CSRF token
      def extract_csrf(html)
        m = html.match(CSRF_META_RE) || html.match(CSRF_INPUT_RE)
        raise "email10min: 未找到 CSRF token" unless m

        m[1]
      end

      # 从 HTML 中提取邮箱地址
      def extract_email(html)
        [EMAIL_ID_RE, EMAIL_DATA_RE, EMAIL_JSON_RE].each do |re|
          if (m = html.match(re))
            addr = m[1].strip
            return addr if addr.include?("@")
          end
        end
        if (m = html.match(EMAIL_GENERIC_RE))
          addr = m[1].strip
          return addr unless addr.include?("email10min") || addr.include?("example")
        end
        raise "email10min: 未找到邮箱地址"
      end

      # 创建临时邮箱
      # @return [EmailInfo]
      def generate_email
        resp = Http.get("#{BASE_URL}/zh", headers: BROWSER_HEADERS)
        resp.raise_for_status
        cookie = merge_cookies(resp)
        html = resp.body
        csrf = extract_csrf(html)
        email = extract_email(html)
        EmailInfo.new(channel: CHANNEL, email: email, token: encode_token(cookie, csrf))
      end

      # 获取邮件列表
      # @param email [String]
      # @param token [String]
      # @return [Array<Email>]
      def get_emails(email, token)
        cookie, csrf = decode_token(token)
        ts = (Time.now.to_f * 1000).to_i
        body = "_token=#{csrf}&captcha="
        resp = Http.post("#{BASE_URL}/messages?#{ts}",
                         headers: AJAX_HEADERS.merge("Cookie" => cookie),
                         body: body)
        resp.raise_for_status
        data = JSON.parse(resp.body)
        messages = data["messages"]
        return [] unless messages.is_a?(Array) && !messages.empty?

        messages.each_with_index.map do |msg, idx|
          id = msg["id"] || msg["message_id"] || idx
          from_addr = msg["from"].to_s
          from_addr = msg["sender"].to_s if from_addr.empty?
          text = msg["text"].to_s
          text = msg["body"].to_s if text.empty?
          html_body = msg["html"].to_s
          html_body = msg["body_html"].to_s if html_body.empty?
          date = msg["date"].to_s
          date = msg["created_at"].to_s if date.empty?
          Normalize.normalize_email({
            "id" => id.to_s,
            "from" => from_addr,
            "to" => msg["to"].to_s.empty? ? email : msg["to"],
            "subject" => msg["subject"].to_s,
            "text" => text,
            "html" => html_body,
            "date" => date,
            "isRead" => false
          }, email)
        end
      end
    end
  end
end
